scipy/stats/_unuran/unuran_wrapper.pyx.templ

```diff
@@ -0,0 +1,1122 @@
+            unur_urng_free(self.urng)
+            self._urng_builder = _URNG(self.numpy_rng)
+            self.urng = self._urng_builder.get_urng()
+            if self.urng == NULL:
+                raise UNURANError(self._messages.get())
+            unur_chg_urng(self.rng, self.urng)
+        finally:
+            _lock.release()
+
+    @cython.final
+    def __dealloc__(self):
+        if self.distr != NULL:
+            unur_distr_free(self.distr)
+            self.distr = NULL
+        if self.par != NULL:
+            unur_par_free(self.par)
+            self.par = NULL
+        if self.rng != NULL:
+            unur_free(self.rng)
+            self.rng = NULL
+        if self.urng != NULL:
+            unur_urng_free(self.urng)
+            self.urng = NULL
+
+    # Pickling support
+    @cython.final
+    def __reduce__(self):
+        klass = functools.partial(self.__class__, **self._kwargs)
+        return (klass, ())
+
+
+cdef class TransformedDensityRejection(Method):
+    r"""
+    Transformed Density Rejection (TDR) Method.
+
+    TDR is an acceptance/rejection method that uses the concavity of a
+    transformed density to construct hat function and squeezes automatically.
+    Most universal algorithms are very slow compared to algorithms that are
+    specialized to that distribution. Algorithms that are fast have a slow
+    setup and require large tables. The aim of this universal method is to
+    provide an algorithm that is not too slow and needs only a short setup.
+    This method can be applied to univariate and unimodal continuous
+    distributions with T-concave density function. See [1]_ and [2]_ for
+    more details.
+
+    Parameters
+    ----------
+    dist : object
+        An instance of a class with 'pdf' and 'dpdf' methods.
+
+        * 'pdf': PDF of the distribution. The signature of the PDF is
+          expected to be: ``def pdf(x: float, *params) -> float``. i.e.
+          the PDF should accept a Python float and the parameters and
+          return a Python float. It doesn't need to integrate to 1 i.e.
+          the PDF doesn't need to be normalized.
+        * 'dpdf': Derivative of the PDF w.r.t x (i.e. the variate). Must
+          have the same signature as the PDF.
+
+    mode : float, optional
+        Mode of the distribution. It is optional and only required if
+        `use_mode` is ``True``.
+    center : float, optional
+        Center (``loc``) of the distribution. It is optional and only
+        required if `use_center` is ``True``. Default is 0.
+    params : tuple, optional
+        Parameters that the PDF and DPDF take. Default is an empty tuple.
+    domain : list or tuple of length 2, optional
+        The support of the distribution. Must not contain `nan` values.
+        Default is ``None``. When ``None``, the support is assumed to be
+        :math:`(-\infty, \infty)`.
+    c : float, optional
+        The transformation function. See Notes for the allowed values for
```

Comment:
If this is a float, is there a more accurate description than "transformation function"?

Reply:
> If this is a float, is there a more accurate description than "transformation function"?

We can copy UNU.RAN's description:

![image](https://user-images.githubusercontent.com/43181252/128586581-ee2fb2f7-40de-4b02-b54e-1370b3e09cab.png)

I think it's much more descriptive.

Reply:
Yes. One thing I was driving at is that it's a parameter of a function; it's not "the transformation function" itself.

Reply:
Done.

Reply:
Now I think I understand a bit better why you called this "The transformation function" originally. You are using the float to select which of two functions to use; `c` is not really used as a continuous "parameter".  It appears that there is an underlying way of describing the transformation function in terms of a parameter `c`, but that is not really done here.

I think we need to decide where this is going to select the most appropriate UI. 

If `c` will someday be a parameter that can actually vary _between_ `0` and `-0.5`, then I think it's OK to leave `c` as a float, describe it as a "parameter" of the transformation function, but document it like:
```
c : {-0.5, 0}, optional
```
and produce an error when it is not one of the allowable values. 

If that is unlikely to happen, then this should be a string e.g.
```
transformation : {"log", "inverse-sqrt"}, optional
```

Reply:
correct, normally c would be a continuous parameter but UNU.RAN just supports 0 and -0.5. I would prefer to keep the name `c` since the transformation is called T_c in the original paper. Should we go with {-0.5, 0}? If someone want to extend the approch to more general c, it could be changed easily

Reply:
> If that is unlikely to happen, then I think this should be a string

I don't think UNU.RAN will add support for other c anytime soon: it hasn't been updated in years! So, this seems the most sensible thing to do. I will change it this way, thanks!

> Should we go with {-0.5, 0}?

Yes but with string name. So, instead of `{-0.5, 0}`, we could have a more descriptive `{'inverse-sqrt', 'log'}`. I hope you agree.

Reply:
After thinking about this a bit more, I agree with @chrisb83 that we should keep parameter name `c` and accept float values `{-0.5, 0.}` (as opposed to {`inverse-sqrt`, `log`}) to be more consistent with the original paper and UNU.RAN's naming convention. If we want, we can later also accept string values while keeping the support for float values intact.